dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java

```diff
@@ -1537,4 +1587,175 @@ public static void main(String[] argv) {
+                    }
+                }
+            }
+        }
+        return newLine;
+    }
+
+    /**
+     * Populates the csvRefMap and csvRowMap for the given csv line.
+     *
+     * The csvRefMap is an index that keeps track of which rows have a specific value for
+     * a specific metadata field or the special "rowName" column. This is used to help resolve indirect
+     * entity references in the same CSV.
+     *
+     * The csvRowMap is a row number to UUID map, and contains an entry for every row that has
+     * been processed so far which has a known (minted) UUID for its item. This is used to help complete
+     * the resolution after the row number has been determined.
+     *
+     * @param line the csv line.
+     * @param rowNumber the row number.
+     * @param uuid the uuid of the item, which may be null if it has not been minted yet.
+     */
+    private void populateRefAndRowMap(DSpaceCSVLine line, int rowNumber, @Nullable UUID uuid) {
+        if (uuid != null) {
+            csvRowMap.put(rowNumber, uuid);
+        }
+        for (String key : line.keys()) {
+            if (key.contains(".") && !key.split("\\.")[0].equalsIgnoreCase("relation") ||
+                    key.equalsIgnoreCase("rowName")) {
+                for (String value : line.get(key)) {
+                    String valueKey = key + ":" + value;
+                    Set<Integer> rowNums = csvRefMap.get(valueKey);
+                    if (rowNums == null) {
+                        rowNums = new HashSet<>();
+                        csvRefMap.put(valueKey, rowNums);
+                    }
+                    rowNums.add(rowNumber);
+                }
+            }
+        }
+    }
+
+    /**
+     * Gets the UUID of the item indicated by the given reference, which may be a direct UUID string, a row reference
+     * of the form rowName:VALUE, or a metadata value reference of the form schema.element[.qualifier]:VALUE.
+     *
+     * The reference may refer to a previously-processed item in the CSV or an item in the database.
+     *
+     * @param context the context to use.
+     * @param reference the reference which may be a UUID, metadata reference, or rowName reference.
+     * @return the uuid.
+     * @throws MetadataImportException if the reference is malformed or ambiguous (refers to multiple items).
+     */
+    private UUID resolveEntityRef(Context context, String reference) throws MetadataImportException {
+        // value reference
+        UUID uuid = null;
+        if (!reference.contains(":")) {
+            // assume it's a UUID
+            try {
+                return UUID.fromString(reference);
+            } catch (IllegalArgumentException e) {
+                throw new MetadataImportException("Not a UUID or indirect entity reference: '" + reference + "'");
+            }
+        } else if (!reference.startsWith("rowName:") ) { //Assume it's a metadata value reference
```

Comment:
@AndrewZWood : This is much improved too.  Thanks for the detailed explanation and inline comments.  My only small comment here is that the error message could provide a bit more information.  Maybe something like:

`"No matches found for reference: " + reference + ".  Keep in mind you can only reference entries that are listed before this one within the CSV."`

The reason why the second sentence may be important is simply because I can anticipate users asking (on mailing lists etc) why the parser here isn't finding a reference that exists in the CSV (as it is possible that there is a matching entry here, but it's just listed *after* the reference).

Reply:
Good point, I've gone ahead and added that improvement here: https://github.com/DSpace/DSpace/pull/2471/commits/c4626c1a4940e202e2062989f5eadbd7c9cfa939